src/polygons.jl

```diff
@@ -775,6 +775,76 @@ function polyfit(plist::Array{Point,1}, npoints = 30)
     return resultpoly
 end
 
+"""
+    bspline(controlPoints::Array{Point,1}, npoints; degree=3, clamped=true)
+
+Generate a B-spline curve from a given set of control points.
+
+# Arguments
+- `controlPoints::Array{Point,1}`: An array of control points that define the B-spline.
+- `npoints=100`: The number of points to generate on the B-spline curve.
+- `degree=3`: The degree of the B-spline. Default is 3.
+- `clamped=true`: A boolean to indicate if the B-spline is clamped. Default is true.
+
+# Returns
+- An array of points on the B-spline curve.
+"""
+function bspline(controlPoints::Array{Point,1}, npoints=30;degree=3, clamped=true)
+    nCP = length(controlPoints)
+    nCP == 0 && error("Error: controlPoints array cannot be empty.")
+    npoints <= 0 && error("Error: npoints must be greater than zero.")
+    degree <= 0 && error("Error: degree must be greater than zero.")
+    degree >= nCP && error("Error: degree cannot be greater than the number of control points.")
+    points = []
+    T = []
+    clamped && (T = fill(0., degree))
+    for i=0:(clamped ? nCP-degree : nCP+degree)
+        push!(T, i / (clamped ? nCP-degree : nCP+degree))
+    end
+    clamped && append!(T, fill(1., degree))
+    
+    """De Boor's algorithm for B-spline evaluation.
+        from https://en.wikipedia.org/wiki/De_Boor%27s_algorithm
+        Arguments
+        ---------
+        k: Index of knot interval that contains x.
+        x: Position.
+        t: Array of knot positions, needs to be padded as described above.
+        c: Array of control points.
+        p: Degree of B-spline.
+    """
+    function deBoor(k,x,t,c,p)
+        d = []
```

Comment:
`d = Float64[]` ?

Reply:
Actually, not necessarily. It can be `d = Point[]` which will satisfy most use cases. If we want to keep it capable of dealing with Float and other arithmetic data types, we'd have to keep it as `[]`

Reply:
I suppose it could be a Union? But I'm guessing the call to this internal function will always be using Points? I know that `Any[]` arrays are not recommended for performance reasons...

Reply:
I'm unsure of whether having a union would provide any performance benefits over an Any[]. I think the most obvious way to maintain generality and performance is to make sure that the dot notation over Point as a tuple. That way the function can take in a Point[] and do both x and y calculations separately. But then again we'd lose out a lot of the contiguous memory accesses and vector operations. Do you know if there a way to make Julia precompile for two different possible data types (i.e. tell it it's either going to be a Float64[] or Point[])?